state/txindex/kv/kv.go

```diff
@@ -40,6 +41,21 @@ type TxIndex struct {
 	log log.Logger
 }
 
+func (txi *TxIndex) Prune(retainHeight int64) {
+	ctx := context.Background()
+	results, err := txi.Search(ctx, query.MustCompile(fmt.Sprintf("tx.height < %d", retainHeight)))
```

Comment:
I think here you can benefit from the same optimization of caching the lastRetain height, so as to avoid calling `Search` for the same height over and over again.